### PR TITLE
Use argparse config throughout app

### PR DESCRIPTION
**Pull request by mplanchard**
This PR is a pretty substantial refactor of the entrypoints of pypiserver (`__main__` and `__init__`) to use the argparse-based config added in #339.

- Updated `RunConfig` and `UpdateConfig` classes to have exclusive init kwargs, instead of taking an namespace. This turned out to be much easier when working with the library-style app initialization in `__init__`, both for direct instantiation and via paste config
- Added an `iter_packages()` method to the `RunConfig` to iterate over packages specified by the configuration (note @elfjes, I think that replacing this with e.g. a `backend` reference will be a nice way to tie in #348)
- Added a general-purpose method to map legacy keyword arguments to the `app()` and `paste_app_factory()` functions to updated forms
- Refactored the `paste_app_factory()` to not mutate the incoming dictionary
- Removed all argument-parsing and config-related code from `__main__` and `core`
- Moved `_logwrite` from `__init__` to `__main__`, since that was the only place it was being used after the updates to `core`
- Updated `digest_file` to use `hashlib.new(algo)` instead of `getattr(hashlib, algo)`, because the former supports more algorithms
- Updated `setup.py` to, instead of calling `eval()` on the entirety of `__init__`, to instead just evaluate the line that defines the version
- Assigned the config to a `._pypiserver_config` attribute on the `Bottle` instance to reduce hacky test workarounds
- Fixed the tox config, which I broke in #339

### CODE REVIEW

**Review by elfjes — COMMENTED** [review]
Looks good, I like it! I have left a few (minor) remarks.

I do have one question regarding the funky logic with importing _app and removing it from the imports again, effectively making `app_from_config` an app-factory. What is the exact requirement for that? Is it just for testing? If so, it may be easier to handle reinitializing the config and other run-time globals in a fixture, and just use the bottle app object as is. Or am I missing something?

File: pypiserver/__init__.py [hunk 1]
```diff
@@ -11,203 +16,172 @@
...
+            f"  {', '.join(replacement_strs)}"
+        )
+        print(warn_str, file=sys.stderr)
+
+    # Create an iterable of 2-tuple to collect into the updated dictionary. Each
+    # item will either be the existing key-value pair from kwargs, or, if the
+    # keyword is a legacy keyword, the new key and potentially adjusted value
+    # for that keyword. Note that depending on the order the argument are
+    # specified, this _could_ mean an updated legacy keyword could override
+    # a new argument if that argument is also specified. For example,
```
[the remark this answers is not in the document]
Comment (mplanchard) [reply pypiserver/__init__.py:84]: Yeah, this is a tricky thing for sure. As we've got it now, it just sort of depends on what order the dictionary is in. I do think it's an edge case, but it also might be better just to throw an error if there are any duplicates. I think we could check that with a pretty simple check to make sure the size of the updated dictionary hasn't changed compared to the size of the incoming dictionary.

File: pypiserver/__init__.py [hunk 2]
```diff
@@ -11,203 +16,172 @@
...
+    # modules map, so that any future imports do not receive our mutated version
     sys.modules.pop("pypiserver._app", None)
-
-    kwds = default_config(**kwds)
-    config, packages = core.configure(**kwds)
     _app.config = config
-    _app.packages = packages
-    _app.app.module = _app  # HACK for testing.
-
+    _app.iter_packages = config.iter_packages
```
[the remark this answers is not in the document]
Comment (mplanchard) [reply pypiserver/__init__.py:117]: Yep, good catch. I put these in here early on and forgot to remove them when I switched everything over to use the config

File: pypiserver/config.py [hunk 3]
```diff
@@ -6,42 +6,55 @@
...
+  the `add_common_args()` function
 - If it should only be available for the `run` command, add it to the
-  `run_parser`.
+  `run_parser` in the `get_parser()` function.
 - If it should only be available for the `update` command, add it to the
-  `update_parser`.
+  `update_parser` in the `get_parser() function`.
 - Add it to the appropriate Config class, `_ConfigCommon` for global options,
   `RunConfig` for `run` options, and `UpdateConfig` for `update` options.
+  - This requires addit it as an `__init__()` kwarg, setting it as an instance
```
[the remark this answers is not in the document]
Comment (mplanchard) [reply pypiserver/config.py:16]: good catch, thanks!

**Comment by mplanchard** [comment]
> I do have one question regarding the funky logic with importing _app and removing it from the imports again, effectively making app_from_config an app-factory. What is the exact requirement for that? Is it just for testing? If so, it may be easier to handle reinitializing the config and other run-time globals in a fixture, and just use the bottle app object as is. 

This is a good question. The way it's done definitely feels funky these days. It's been that way since long before I started working on the project, and it's never been a big enough pain to bother changing. The reason it is the way it is, I'm pretty sure, is that we're setting global references in our `_app` reference to the config and so on. If someone call the `app()` constructor twice, we want them to get a new version of that import with whatever the newly generated config is, _without_ changing the config and globals on the old reference. I'll think on whether there's an easy way to update this, if not with this PR, then potentially with another. Ideally I think we would just provide this stuff via some kind of `context` object that comes in with the bottle arguments.

**Comment by mplanchard** [comment]
Going to merge this so we can go ahead and start working on whatever painful merges it's going to require for #348 :sweat_smile: 

Let me know if you would like help with that! I can always open a PR against your branch.

**Comment by mplanchard** [comment]
> I do have one question regarding the funky logic with importing _app and removing it from the imports again, effectively making app_from_config an app-factory.

Opened #350 to address this!

**Pull request merged**